src/ModelContextProtocol/McpChatClientBuilderExtensions.cs

```diff
@@ -0,0 +1,230 @@
+                        case HostedMcpServerToolNeverRequireApprovalMode neverRequireApproval:
+                            downstreamTools.Add(mcpFunction);
+                            break;
+                        case HostedMcpServerToolRequireSpecificApprovalMode specificApprovalMode when specificApprovalMode.AlwaysRequireApprovalToolNames?.Contains(mcpFunction.Name) is true:
+                            downstreamTools.Add(new ApprovalRequiredAIFunction(mcpFunction));
+                            break;
+                        case HostedMcpServerToolRequireSpecificApprovalMode specificApprovalMode when specificApprovalMode.NeverRequireApprovalToolNames?.Contains(mcpFunction.Name) is true:
+                            downstreamTools.Add(mcpFunction);
+                            break;
+                        default:
+                            // Default to always require approval if no specific mode is set.
+                            downstreamTools.Add(new ApprovalRequiredAIFunction(mcpFunction));
+                            break;
+                    }
+                }
+            }
+
+            return downstreamTools;
+        }
+
+        /// <inheritdoc/>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                // Dispose of the HTTP client if it was created by this client.
+                if (_ownsHttpClient)
+                {
+                    _httpClient?.Dispose();
+                }
+
+                if (_mcpClientTasks is not null)
+                {
+                    // Dispose of all cached MCP clients.
+                    foreach (var clientTask in _mcpClientTasks.Values)
+                    {
+#if NETSTANDARD2_0
+                        if (clientTask.Status == TaskStatus.RanToCompletion)
+#else
+                        if (clientTask.IsCompletedSuccessfully)
+#endif
+                        {
+                            _ = clientTask.Result.DisposeAsync();
+                        }
+                    }
+
+                    _mcpClientTasks.Clear();
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        private Task<McpClient> CreateMcpClientAsync(Uri serverAddress, string serverName, string? authorizationToken)
+        {
+            if (_mcpClientTasks is null)
+            {
+                _mcpClientTasks = new ConcurrentDictionary<string, Task<McpClient>>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            // Note: We don't pass cancellationToken to the factory because the cached task should not be tied to any single caller's cancellation token.
+            // Instead, callers can cancel waiting for the task, but the connection attempt itself will complete independently.
+            return _mcpClientTasks.GetOrAdd(serverAddress.ToString(), _ => CreateMcpClientCoreAsync(serverAddress, serverName, authorizationToken, CancellationToken.None));
```

Comment:
Probably not a huge deal given the other overheads, but this is always going to allocate a closure to capture serverAddress, serverName, and authorizationToken.